packages/react-storage/src/components/StorageBrowser/Views/LocationActionView/useHandleUpload.ts

```diff
@@ -129,60 +152,148 @@ export function useHandleUpload({
+    // has status 'CANCELED' and its 'cancel' property is undefined
     setTasks((prevTasks) =>
-      prevTasks.map(({ key, data, progress, size }) => {
-        const { bucket: bucketName, credentialsProvider, region } = getConfig();
-        const input: UploadDataWithPathInput = {
-          path: `${prefix}${key}`,
-          data,
-          options: {
-            bucket: { bucketName, region },
-            locationCredentialsProvider: credentialsProvider,
-            onProgress: ({ totalBytes, transferredBytes }) => {
-              const progress = totalBytes ? transferredBytes / totalBytes : 0;
-              const nextTask: TaskUpdate = {
-                key,
-                progress,
-                ...(totalBytes === transferredBytes
-                  ? { cancel: undefined }
-                  : undefined),
-              };
-              setTasks((curr) => updateTasks(curr, nextTask));
-            },
-            preventOverwrite,
-          },
-        };
+      prevTasks.map((task) => ({
+        ...task,
+        status: 'QUEUED',
+        cancel: () =>
+          setTasks((prevTasks) =>
+            updateTasks(prevTasks, {
+              key: task.key,
+              status: 'CANCELED',
+              cancel: undefined,
+            })
+          ),
+      }))
+    );
 
-        const { cancel, result } = uploadData(input);
+    const uploadNext = async () => {
+      const currentTasks = tasksRef.current;
 
-        result
-          .then(() => {
-            setTasks((curr) => updateTasks(curr, { key, status: 'COMPLETE' }));
+      if (currentIndex >= currentTasks.length) {
+        return;
+      }
+
+      const queuedTask = currentTasks[currentIndex];
+
+      if (queuedTask.status === 'CANCELED') {
+        currentIndex++;
+
+        uploadNext();
+
+        return;
+      }
+
+      const pendingTask = processUpload(queuedTask);
+
+      currentIndex++;
+
+      const { task, result } = pendingTask;
+      const { key } = task;
+
+      try {
+        await result;
+
+        setTasks((prevTasks) =>
+          updateTasks(prevTasks, { key, status: 'COMPLETE' })
+        );
+      } catch (error) {
+        setTasks((prevTasks) =>
+          updateTasks(prevTasks, {
+            key,
+            status: isCancelError(error) ? 'CANCELED' : 'FAILED',
+            cancel: undefined,
           })
-          .catch(() => {
-            setTasks((curr) => updateTasks(curr, { key, status: 'FAILED' }));
-          });
+        );
+      } finally {
+        uploadNext();
+      }
+    };
 
-        const handleCancel = () => {
-          cancel();
+    // Start the initial set of uploads
+    const initialUploads = Array(batchSize)
+      .fill(null)
+      .map(() => uploadNext());
 
-          setTasks((curr) =>
-            // assign cancel to noop
-            updateTasks(curr, { key, cancel: () => null, status: 'CANCELED' })
-          );
-        };
+    await Promise.all(initialUploads);
```

Comment:
There is no reason to use `Promise.all` unless it is for tracking whether all tasks have completed or an exception has been reached. There is some nuance with the concurrency behavior we are looking for in this hook as it should start the next task as soon as a task completes, not once a btach completes.

Additionally believe that `Promise.all` would introduce bugs since it resolves on any exception, meaning that if task 1 of 6 throws then the `Promise.all` would reject while the other 5 tasks are still running. If awaiting a batch was the desired behavior, `Promise.allSettled` would be what we are looking for here